### PR TITLE
Install a bottle from an URL

### PR DESCRIPTION
**Pull request by sjackman**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/homebrew/pull/49031).
- [x] Have you successfully run `brew tests` with your changes locally?

-----

This PR makes it possible to install a bottle from an URL. For example
```sh
brew install https://4574-56942570-gh.circle-artifacts.com/0/tmp/circle-artifacts.PaJXYOU/git-2.12.2.x86_64_linux.bottle.tar.gz
```

### CODE REVIEW

**Comment by chdiza** [comment]
This sounds great, particularly if the URL can be of the form `file:///blah/bloo/blee`.

**Comment by MikeMcQuaid** [comment]
@chdiza We already support installing bottles from file paths, FYI.

**Comment by chdiza** [comment]
> We already support installing bottles from file paths, FYI.

I haven't tried in a while, but I found this to be pretty buggy last time (at least when the bottles were "bottled" locally, rather than being cached from Bintray).

**Comment by sjackman** [comment]
> I haven't tried in a while, but I found this to be pretty buggy last time (at least when the bottles were "bottled" locally, rather than being cached from Bintray).

It works for me with two caveats.
1. The bottle version has to match the formula in the repo.
2. The formula in the repo must have a `sha256` for this OS.

1. is by design, but could be relaxed now that that the bottle includes the formula inside the bottle.
2. is a bug in my opinion.

**Comment by sjackman** [comment]
TravisCI on macOS failed. It passed on Linux.
```
==> brew update-test --to-tag
Error: undefined method `chomp' for nil:NilClass
Please report this bug:
  http://docs.brew.sh/Troubleshooting.html
/Users/travis/build/Homebrew/brew/Library/Homebrew/dev-cmd/update-test.rb:36:in `update_test'
/Users/travis/build/Homebrew/brew/Library/Homebrew/brew.rb:91:in `<main>'
==> FAILED
```
https://travis-ci.org/Homebrew/brew/jobs/215813460#L177

This bug was hopefully (fingers crossed) resolved by https://github.com/Homebrew/brew/pull/2404
I'll rebase onto `master`.

**Comment by sjackman** [comment]
The TravisCI failure is resolved after rebasing onto `master`.

File: Library/Homebrew/formulary.rb [hunk 1]
```diff
@@ -105,7 +105,16 @@ def load_file
   # Loads formulae from bottles.
   class BottleLoader < FormulaLoader
     def initialize(bottle_name)
-      @bottle_filename = Pathname(bottle_name).realpath
+      case bottle_name
+      when %r{(https?|ftp|file)://}
+        @bottle_filename = HOMEBREW_CACHE/File.basename(bottle_name)
+        unless @bottle_filename.exist?
+          HOMEBREW_CACHE.mkpath
+          curl bottle_name, "-o", @bottle_filename
```
Comment (MikeMcQuaid) [inline Library/Homebrew/formulary.rb]: I think the `CurlBottleDownloadStrategy` should be used here instead to handle e.g. caching and resuming.

**Comment by chdiza** [comment]
> We already support installing bottles from file paths

Tried this again just now, to pour a locally-held bottle that I myself built from a custom formula in my own tap.  (I.e., using `--build-bottle`, it is not just the bintray bottle held locally).  I get:

```
$ brew install file:///Users/chdiza/gnu-tar-1.29.sierra.bottle.tar.gz
[curl is doing its thing]
Error: gnu-tar-1.29.sierra.bottle.tar.gz: /stuff/brache/Formula/gnu-tar-1.29.sierra.bottle.tar.gz:1: Invalid char `\x1F' in expression
/stuff/brache/Formula/gnu-tar-1.29.sierra.bottle.tar.gz:1: invalid multibyte char (UTF-8)
==> Searching for similarly named formulae...
Error: No similarly named formulae found.
==> Searching taps...
Error: No formulae found in taps.
```
I get that same error when trying other packages too.

**Comment by sjackman** [comment]
With this PR, the above command should work. Was your test with or without this PR applied?

Without this PR applied, this works:
```sh
brew install /Users/chdiza/gnu-tar-1.29.sierra.bottle.tar.gz
```
and this does not work:
```sh
brew install file:///Users/chdiza/gnu-tar-1.29.sierra.bottle.tar.gz
```

File: Library/Homebrew/formulary.rb (hunk 1 above)
Comment (sjackman) [inline Library/Homebrew/formulary.rb]: Fixed. Thanks for the helpful suggestion, Mike.

**Comment by chdiza** [comment]
> Was your test with or without this PR applied?

It was without.

> brew install /Users/chdiza/gnu-tar-1.29.sierra.bottle.tar.gz

That does NOT work sans the PR for me.  It merely tries building the package from source.

I will get to testing the PR momentarily.

**Comment by sjackman** [comment]
> That does NOT work sans the PR for me. It merely tries building the package from source.

I'm not sure that this PR will help that. All it does is download the bottle to cache, and then install that file.

> I will get to testing the PR momentarily.

Thanks!

**Comment by chdiza** [comment]
It also doesn't work even with the PR.  It still tries to build from source no matter what:
```
brew install --force-bottle  file:///users/chdiza/gzip-1.8.sierra.bottle.tar.gz
==> Downloading file:///users/chdiza/gzip-1.8.sierra.bottle.tar.gz
Already downloaded: /stuff/brache/gzip-1.8.sierra.bottle.tar.gz
Warning: Pouring the cached bottle.
==> Installing gzip from chdiza/dizaone
==> Downloading file:///stuff/bb/unix/gzip-1.8.tar.xz
Already downloaded: /stuff/brache/gzip-1.8.tar.xz
==> Verifying gzip-1.8.tar.xz checksum
tar xJf /stuff/brache/gzip-1.8.tar.xz
==> ./configure --prefix=/usr/local/Cellar/gzip/1.8
```

However, *with* the PR at least I'm not getting the "invalid multibyte char" error from above.

**Comment by sjackman** [comment]
The formula for `gzip` doesn't have a bottle for `sierra`. See my comment above
https://github.com/Homebrew/brew/pull/2406#issuecomment-289825278

**Comment by chdiza** [comment]
> The formula for gzip doesn't have a bottle sha256 for sierra. See my comment above
#2406 (comment)

OK.  So pouring local bottles that are made from one's custom tap are actually not supported.  I had thought that the only thing preventing this was that the CLI didn't seem to like a URL or filepath-to-the-gzipped-bottle as the argument.  So I thought that this PR would fix that bit, and then I could pour custom bottles.  I was wrong.

**Comment by sjackman** [comment]
It's a bug in my opinion, and shouldn't be too terribly hard to fix. When pouring a local bottle or URL, it needs to skip the check for a `sha256` in the formula.

**Comment by sjackman** [comment]
Note that the `sha256` in the formula doesn't have to match that of the local bottle, it just has to exist at all. `sha256 "" => :sierra` is good enough.

**Comment by chdiza** [comment]
> Note that the bottle sha256 in the formula doesn't have to match that of the local bottle, it just has to exist. sha256 "" => :sierra is good enough.

I don't think this is true.  `gnu-tar` has a sha256 for sierra, but I can't pour it from a path or URL either, with or without this PR.  (Same result: `brew` just tries to build it from source.)

**Comment by sjackman** [comment]
```
❯❯❯ brew install ~/Library/Caches/Homebrew/gnu-tar-1.29_1.el_capitan.bottle.tar.gz
==> Pouring gnu-tar-1.29_1.el_capitan.bottle.tar.gz
```
I'm not able to replicate that behaviour on `:el_capitan`.

**Comment by chdiza** [comment]
> I'm not able to replicate that behaviour on :el_capitan.

Is the bottle you're trying to install just a cached version of the Bintray bottle?  I'm trying to install a bottle built from my own `gnu-tar` formula in my own tap.

**Comment by sjackman** [comment]
Yes. Ah, I see. Bottling works for our custom taps `Linuxbrew/extra` and friends, but you may have some debugging to do. To eliminate one source of possible confusion, you may want to first try it on a formula name that's not also found in `Homebrew/core`.

**Comment by sjackman** [comment]
@MikeMcQuaid Good to merge?

File: Library/Homebrew/formulary.rb [hunk 2]
```diff
@@ -105,7 +105,17 @@ def load_file
   # Loads formulae from bottles.
   class BottleLoader < FormulaLoader
     def initialize(bottle_name)
-      @bottle_filename = Pathname(bottle_name).realpath
+      case bottle_name
+      when %r{(https?|ftp|file)://}
+        resource = Resource.new bottle_name[%r{([^/]+)-}, 1] { url bottle_name }
```
Comment (MikeMcQuaid) [inline Library/Homebrew/formulary.rb]: Not sure I understand the regex here, maybe warrants a comment?

File: Library/Homebrew/formulary.rb [hunk 3]
```diff
@@ -105,7 +105,17 @@ def load_file
...
+        downloader = CurlBottleDownloadStrategy.new resource.name, resource
+        @bottle_filename = downloader.cached_location
+        cached = @bottle_filename.exist?
+        downloader.fetch
+        opoo "Pouring the cached bottle." if cached
```
Comment (MikeMcQuaid) [inline Library/Homebrew/formulary.rb]: `ohai` and skip the trailing `.`

**Comment by MikeMcQuaid** [comment]
@chdiza Does this work for you now?

**Comment by chdiza** [comment]
> @chdiza Does this work for you now?

I won't be in a position to check whether it works for the next several hours, at least.  Probably just merge it and I'll open a separate issue if needed.

File: Library/Homebrew/formulary.rb (hunk 2 above)
Comment (sjackman) [inline Library/Homebrew/formulary.rb]: Done. The name of the formula is found between the last slash and the last hyphen.

File: Library/Homebrew/formulary.rb (hunk 3 above)
Comment (sjackman) [inline Library/Homebrew/formulary.rb]: Done.

**Comment by sjackman** [comment]
Thanks for the feedback, Mike. Updated.

**Comment by MikeMcQuaid** [comment]
Thanks @sjackman!

**Comment by sjackman** [comment]
My pleasure. Thanks for merging, Mike.

**Pull request merged**